### PR TITLE
Fix componentpath formrelativepath

### PR DESCRIPTION
**Pull request by adamkewley**
This fixes a bug found here:

https://github.com/opensim-org/opensim-core/pull/2805

It's explained in more detail on that PR but, effectively, `ComponentPath::formRelativePath` is broken on the edge-case of a path subcomponent containing a non-separator at the same offset as another path subcomponent containing its NUL terminator (e.g. `/component`, `/component_path`).

---
This change is [Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2890)

### CODE REVIEW

**Comment by adamkewley** [comment]
The other edge-cases I need to double-check are whether the old implementation returns `""` or `"."` when both paths are identical. Logically, it should be "." (`""` is not a path), but I have a suspicion that it returns `""`.

**Comment by adamkewley** [comment]
This should be basically ready for review, provided CI passes.

**Review by nickbianco — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2890)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @adamkewley)

___
*[OpenSim/Common/Test/testPath.cpp, line 120 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/2890#-MJ8gCZlDeAtteWDX4K7:-MJ8gCZlDeAtteWDX4K8:b-fhqt0z) ([raw file](https://github.com/opensim-org/opensim-core/blob/9a4a4b46d6c8de56e9dc212e151c587294496efd/OpenSim/Common/Test/testPath.cpp#L120)):*
> ```cpp
>                 ss << "  expected = " << tc.expected << std::endl;
>                 ss << "       got = " << ans.toString();
>                 throw std::runtime_error{ss.str()};
> ```

We should use the new logging tools here instead (i.e., `log_error()`).

**Review by adamkewley — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2890)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @nickbianco)

___
*[OpenSim/Common/Test/testPath.cpp, line 120 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/2890#-MJ8gCZlDeAtteWDX4K7:-MJ8tuCX96vcZzteio3m:bvkfbc4) ([raw file](https://github.com/opensim-org/opensim-core/blob/9a4a4b46d6c8de56e9dc212e151c587294496efd/OpenSim/Common/Test/testPath.cpp#L120)):*
<details><summary><i>Previously, nickbianco (Nicholas Bianco) wrote…</i></summary><blockquote>

We should use the new logging tools here instead (i.e., `log_error()`).
</blockquote></details>

Assertion errors aren't typically logged in test suites unless its a stateful test that needs to log a bunch of info just before it fails. So I don't think logging is the right fit. Assuming there's a `format` equivalent, though, this could be rewritten to as:

```c++
            throw std::runtime_error{format(R"(ComponentPath::formRelativePath produced an invalid output
      from = {}
        to = {}
  expected = {}
       got = {}
)", tc.from, tc.to, tc.expected, tc.got)};
```

Not entirely sure if it's worth the effort of an edit, because the old form is pretty standard C++, and isn't particularly ugly. If you're insistent, I'll do it once I get around to a second pass

**Review by nickbianco — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2890)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @adamkewley)

___
*[OpenSim/Common/Test/testPath.cpp, line 99 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/2890#-MJ9EZe46h0EJiQPDdat:-MJ9EZe46h0EJiQPDdau:b32v57) ([raw file](https://github.com/opensim-org/opensim-core/blob/9a4a4b46d6c8de56e9dc212e151c587294496efd/OpenSim/Common/Test/testPath.cpp#L99)):*
> ```cpp
>             // An example that failed post-merge
>             //     see https://github.com/opensim-org/opensim-core/pull/2805
>             { "/contact", "/contact_point", "../contact_point" },
> ```

+1
___
*[OpenSim/Common/Test/testPath.cpp, line 120 at r2](https://reviewable.io/reviews/opensim-org/opensim-core/2890#-MJ8gCZlDeAtteWDX4K7:-MJ9Br0ODpUAGKy8_JVE:bkx5t45) ([raw file](https://github.com/opensim-org/opensim-core/blob/9a4a4b46d6c8de56e9dc212e151c587294496efd/OpenSim/Common/Test/testPath.cpp#L120)):*
<details><summary><i>Previously, adamkewley (Adam Kewley) wrote…</i></summary><blockquote>

Assertion errors aren't typically logged in test suites unless its a stateful test that needs to log a bunch of info just before it fails. So I don't think logging is the right fit. Assuming there's a `format` equivalent, though, this could be rewritten to as:

```c++
            throw std::runtime_error{format(R"(ComponentPath::formRelativePath produced an invalid output
      from = {}
        to = {}
  expected = {}
       got = {}
)", tc.from, tc.to, tc.expected, tc.got)};
```

Not entirely sure if it's worth the effort of an edit, because the old form is pretty standard C++, and isn't particularly ugly. If you're insistent, I'll do it once I get around to a second pass
</blockquote></details>

Oh yes, thanks for the clarification. It's fine as is.

**Comment by adamkewley** [comment]
Going to merge this: it's a non-api-breaking fix that is necessary for the Moco merge.

**Comment by nickbianco** [comment]
Hey @adamkewley, thanks for creating the fix. We usually wait for an approved review before merging. I left some approving comments, but not a full review. Sorry if there was some confusion there.

@aymanhab is going to take a look over the changes since he's more familiar with the recent ComponentPath changes.

**Comment by aymanhab** [comment]
LGTM @adamkewley just had a question regarding possible end-cases before we close the door on this refactoring/fix. Thank you.

**Pull request merged**
